src/fheroes2/game/game_startgame.cpp

```diff
@@ -335,6 +335,8 @@ void Game::OpenCastleDialog( Castle & castle, bool updateFocus /* = true */, con
 
         // The castle garrison can change
         adventureMapInterface.RedrawFocus();
+        // A new hero could be hired in the castle
+        adventureMapInterface.ResetFocus( Interface::GetFocusType(), false );
```

Comment:
Hi @oleg-derevenetz, to fix this issue and keep https://github.com/ihhub/fheroes2/issues/9522 fixed we can add a condition:
```suggestion
        if ( heroCountBefore != myKingdom.GetHeroes().size() ) {
            // A new hero could be hired in the castle
            adventureMapInterface.ResetFocus( Interface::GetFocusType(), false );
        }
```

What do you think?

Reply:
> Hi @Districh-ru Need to figure this out a bit, because there is also some focus-related logic before these lines, inside the `if ( renderBackgroundDialog ) { ... }`, maybe all this can be simplified somehow... If you want, you are free to take a look, especially since you wrote this part, as far as I can see.

@oleg-derevenetz, I'll take a look and push a commit to this PR after doing some tests.

Reply:
@oleg-derevenetz, @zenseii,
I've updated the logic and checked hero recruiting: from the current castle, when switching between castles, when castle is opened from the Kingdom overview; checked hero dismissing and recruiting a hero from the jail.
#9522 is also fixed.

Could you please check the changes I made.

Reply:
Hi @Districh-ru looks good to me, I tried different quirks by nested switching between castles and heroes and dismissing/hiring them and everything worked correctly.